keps/sig-storage/20190117-SkipVolumeOwnershipChange.md

```diff
@@ -0,0 +1,147 @@
+
+### Goals
+
+ - Allow volume ownership and permission to be skipped during mount
+
+### Non-Goals
+
+ - In some cases if user brings in a large enough volume from outside, the first time ownership and permission change still could take lot of time.
+ - On SELinux enabled distributions we will still do recursive chcon whenever applicable and handling that is outside the scope.
+
+## Proposal
+
+We propose that an user can optionally opt-in to skip recursive ownership(and permission) change on the volume if volume already has right permissions.
+
+### Implementation Details/Notes/Constraints [optional]
+
+When creating a pod, we propose that `PersistentVolumeClaimVolumeSource` field expanded to include a new field called `PermissionChangePolicy` which can have following possible values:
+
+ - `NoChange` --> Don't change permissions and ownership.
+ - `Always` --> Always change the permissions and ownership to match fsGroup. This is the current behavior and it will be the default one when this proposal is implemented.
+ - `OnDemand` --> Only change permissions when fsGroup of pod and PVC don't match.
+
+```go
+type PersistentVolumeClaimVolumeSource struct {
+    // ClaimName is the name of a PersistentVolumeClaim in the same namespace as the pod using this volume
+    ClaimName string
+    // Optional: Defaults to false (read/write).  ReadOnly here
+    // will force the ReadOnly setting in VolumeMounts
+    // +optional
+    ReadOnly bool
+    // PermissionChangePolicy ← new field
+    PermissionChangePolicy string
+}
+```
+
+In addition to this, we propose a new field to PVC's status:
+
+```go
+type PersistentVolumeClaimStatus struct {
+     // Phase represents the current phase of PersistentVolumeClaim
+     // +optional
+     Phase PersistentVolumeClaimPhase
+     // AccessModes contains all ways the volume backing the PVC can be mounted
+     // +optional
+     AccessModes []PersistentVolumeAccessMode
+     // Represents the actual resources of the underlying volume
+     // +optional
+     Capacity ResourceList
+     // FSGroup of PVC
+     // + optional
+     FSGroup *int64     // <------ NEW ------
+     // +optional
+     Conditions []PersistentVolumeClaimCondition
+}
+```
+
+
+When a volume is mounted by the kubelet, volumemanager will check `pvc.Status.FSGroup` and `pod.SecurityContext.FSGroup` and depending on 
+`PermissionChangePolicy` of `PersistentVolumeClaimVolumeSource` it will:
+
+ - Do no permission and ownership change if `PermissionChangePolicy` is `NoChange`.
+ - Will change ownership and permission of volume to match `pod.SecurityContext.FSGroup` if `PermissionChangePolicy` is set to `Always`.
+ - if `PermissionChangePolicy` is set to `OnDemand`, volume permissions will ONLY be changed if `pvc.Status.FSGroup` and `pod.SecurityContext.FSGroup` don't have the same value.
+
+After permissions are changed, pvc.Status.FSGroup will be updated to reflect latest value.
```

Comment:
Ahh, I think I see.  A new claim would have fsgroup = nil.  After one chmod, it would be set, thenceforth faster.  OK, I think that is OK.

One issue - this requires kubelet to write to PVC status -- is that a net-new permission?

@liggitt

Reply:
kubelets have narrow PVC status update permissions if the ExpandPersistentVolumes feature is enabled (only status, and only the capacity/conditions fields currently). that could be expanded to a new field for a new feature

Reply:
okay - so I imagine this has to be included as `or` on existing feature gate check or do we want to allow `pvc.Status` update in general when pvc and node relationship can be established?